pkg/vm/engine/txn/shard_hash.go

```diff
@@ -0,0 +1,596 @@
+			continue
+		}
+		sharded = append(sharded, &ShardedBatch{
+			Shard: *shard,
+			Batch: bat,
+		})
+	}
+
+	return
+}
+
+func (h *HashShard) Vector(
+	ctx context.Context,
+	tableID string,
+	getDefs getDefsFunc,
+	colName string,
+	vec *vector.Vector,
+	nodes []logservicepb.DNStore,
+) (
+	sharded []*ShardedVector,
+	err error,
+) {
+
+	// get defs
+	defs, err := getDefs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// get shard key
+	var shardAttr *engine.Attribute
+	for _, def := range defs {
+		attr, ok := def.(*engine.AttributeDef)
+		if !ok {
+			continue
+		}
+		if attr.Attr.Primary {
+			if attr.Attr.Name == colName {
+				shardAttr = &attr.Attr
+				break
+			}
+		}
+	}
+	if shardAttr == nil {
+		// no shard key
+		return nil, nil
+	}
+
+	// shards
+	var shards []*Shard
+	for _, store := range nodes {
+		for _, info := range store.Shards {
+			shards = append(shards, &Shard{
+				DNShardRecord: metadata.DNShardRecord{
+					ShardID: info.ShardID,
+				},
+				ReplicaID: info.ReplicaID,
+				Address:   store.ServiceAddress,
+			})
+		}
+	}
+	sort.Slice(shards, func(i, j int) bool {
+		return shards[i].DNShardRecord.ShardID < shards[j].DNShardRecord.ShardID
+	})
+	m := make(map[*Shard]*vector.Vector)
+
+	// shard vector
+	for i := 0; i < vec.Length(); i++ {
+		hasher := fnv.New32()
+		bs, err := getBytesFromPrimaryVectorForHash(vec, i, shardAttr.Type)
+		if err != nil {
+			return nil, err
+		}
+		_, err = hasher.Write(bs)
+		if err != nil {
+			panic(err)
+		}
+		n := int(hasher.Sum32())
+		shard := shards[n%len(shards)]
+		shardVec, ok := m[shard]
+		if !ok {
+			shardVec = vector.New(shardAttr.Type)
+			m[shard] = shardVec
+		}
+		v := getNullableValueFromVector(vec, i)
+		appendNullableValueToVector(shardVec, v, h.heap)
+	}
+
+	for shard, vec := range m {
+		if vec.Length() == 0 {
+			continue
+		}
+		sharded = append(sharded, &ShardedVector{
+			Shard:  *shard,
+			Vector: vec,
+		})
+	}
+
+	return
+}
+
+var _ ShardPolicy = new(HashShard)
+
+func (h *HashShard) Stores(stores []logservicepb.DNStore) (shards []metadata.DNShard, err error) {
+	for _, store := range stores {
+		info := store.Shards[0]
+		shards = append(shards, Shard{
+			DNShardRecord: metadata.DNShardRecord{
+				ShardID: info.ShardID,
+			},
+			ReplicaID: info.ReplicaID,
+			Address:   store.ServiceAddress,
+		})
+	}
+	return
+}
+
+func getBytesFromPrimaryVectorForHash(vec *vector.Vector, i int, typ types.Type) ([]byte, error) {
+	if vec.IsConst() {
+		panic("primary value vector should not be const")
+	}
+	if vec.GetNulls().Any() {
+		//TODO mimic to pass BVT
+		return nil, taedata.ErrDuplicate
+		//panic("primary value vector should not contain nulls")
+	}
+	if vec.Typ.Oid.FixedLength() > 0 {
+		// is slice
+		size := vec.Typ.TypeSize()
+		l := vec.Length() * size
+		data := unsafe.Slice((*byte)(vector.GetPtrAt(vec, 0)), l)
+		return data[i*size : (i+1)*size], nil
+	}
+	return vec.GetBytes(int64(i)), nil
+}
+
+type Nullable struct {
+	IsNull bool
+	Value  any
```

Comment:
Do you have to use this Nullable?    Why don't just check vector?

That is , work on whole vector, using nulls and MustTCol.

Reply:
The mem engine is row-based and rows need to be updated individually. I don't think vectors are suitable in row-based structures.

Reply:
I know what you mean. I'll change to use nulls to mask elements.